### PR TITLE
DAOS-15617 test: Find mpi4py test in /usr/bin

### PR DESCRIPTION
**Pull request by brianjmurrell**
### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by daltonbohning — APPROVED** [review]
This is a workaround and not a generic solution, but FWIW we do the same for some HDF5 tests

**Comment by brianjmurrell** [comment]
> This is a workaround and not a generic solution,

Can you expand?  I am intending this to be the final solution, not a work-around.  That is, the tests live in `/usr/bin/` instead of in the Python `site-packages` dir.

>  but FWIW we do the same for some HDF5 tests

Can you point me in the direction of that so that I can see how these compare?

**Comment by daltonbohning** [comment]
> > This is a workaround and not a generic solution,
> 
> Can you expand? I am intending this to be the final solution, not a work-around. That is, the tests live in `/usr/bin/` instead of in the Python `site-packages` dir.

We can't assume that's always where the binary will be. For OUR CI packaging, sure. But what if someone installs a different version to a different location? E.g. if there are multiple python versions installed with binaries in different paths. We don't currently run functional tests in a python virtualenv but IMO _**we really should be**_


> > but FWIW we do the same for some HDF5 tests
> 
> Can you point me in the direction of that so that I can see how these compare?


E.g. https://github.com/daos-stack/daos/blob/51f4fcaa5b4730eebe3b5d1eb385f9ccbf222c5b/src/tests/ftest/ior/small.yaml#L68
If a different MPI is loaded, that path is very likely incorrect.

> FWIW, I do intend to push a more final version of this PR but just want to solicit opinions on the solution before I push anything more final.

A more complete solution should probably look in the env/PATH, sitepackages, etc.

**Comment by brianjmurrell** [comment]
> We can't assume that's always where the binary will be. For OUR CI packaging, sure.

In this particular case, our own packaging is required as we customize (in fact we create) the test in the RPM packaging.

> But what if someone installs a different version to a different location?

Anyone installing an mpi4py RPM from anywhere else will not even get the test we run included in it.  We probably ought to have our mpi4py `Provides: mpi4py-daos` and that should be our dependency name.  But that's another story.

> E.g. if there are multiple python versions installed with binaries in different paths. We don't currently run functional tests in a python virtualenv but IMO _**we really should be**_

Mixing of packaging tools (RPM vs. Pypi/venv/etc.) is what has caused this mess.

> https://github.com/daos-stack/daos/blob/51f4fcaa5b4730eebe3b5d1eb385f9ccbf222c5b/src/tests/ftest/ior/small.yaml#L68

But that is correct, for mpich.
  
> If a different MPI is loaded, that path is very likely incorrect.

Agreed.  This is just poor MPI support in our test framework.  Ideally our framework could use any MPI (specified as an option/configuration) and such a comprehensive MPI-supporting implementation would deal with that hard-coded `mpich`.

> A more complete solution should probably look in the env/PATH, sitepackages, etc.

That is what was being done (and was working) but due to the mixing of the system (RPM) Python and a _venv_ Python, this broke and caused DAOS-15617.

**Comment by daltonbohning** [comment]
> Anyone installing an mpi4py RPM from anywhere else will not even get the test we run included in it. We probably ought to have our mpi4py `Provides: mpi4py-daos` and that should be our dependency name. But that's another story.

This might be okay, but technically it means this test wouldn't work outside our CI environment.


> > A more complete solution should probably look in the env/PATH, sitepackages, etc.
> 
> That is what was being done (and was working) but due to the mixing of the system (RPM) Python and a _venv_ Python, this broke and caused DAOS-15617.

But we don't want to use the system python. AFAIK, the default is python3.6 on all our distros and 3.6 was EOL _**2.5 years ago**_.  We can't keep supporting legacy versions. We could install a newer python RPM, but there aren't python packages available for every dependency we need. And even if there were, when we install `pydaos`, it's packaged/built with the system (or arbitrary) python version. I wonder if we should actually be packaging something like `python3.X-pydaos`. And in fact, it should probably be a separate project/repo.

**Comment by ashleypittman** [comment]
Two comments here:

1. `/usr/bin/` probably isn't the first choice for test binaries.  These are never going to call by a user so not on `PATH` would be nice but I can't immediately say a location that would be better.  Perhaps `/usr/share/mpi4py-daos/`?
2. Some level of interoperability would be nice here or this is going to cause test failures.  Perhaps this code could keep checking the existing path and add the new one, then after a seven day window we could land the packaging change and PRs that were less than a week old would not see test failures at any point.  Maybe if this isn't a default test then we don't need to be as careful.

**Comment by daltonbohning** [comment]
> 2\. Some level of interoperability would be nice here or this is going to cause test failures.  Perhaps this code could keep checking the existing path and add the new one, then after a seven day window we could land the packaging change and PRs that were less than a week old would not see test failures at any point.  Maybe if this isn't a default test then we don't need to be as careful.

Given this test is failing in master and it's a daily (not pr) I don't think we need to be so careful

**Comment by ashleypittman** [comment]
> > 2. Some level of interoperability would be nice here or this is going to cause test failures.  Perhaps this code could keep checking the existing path and add the new one, then after a seven day window we could land the packaging change and PRs that were less than a week old would not see test failures at any point.  Maybe if this isn't a default test then we don't need to be as careful.
> 
> Given this test is failing in master and it's a daily (not pr) I don't think we need to be so careful

Is this test failing in master?  It would only be doing that if we'd landed the packaging change already I'd expect.

**Comment by brianjmurrell** [comment]
> This might be okay, but technically it means this test wouldn't work outside our CI environment.

Yes.  This but a raindrop in an ocean of problems with our testing working outside of our CI environment.  Although, maybe it's not, entirely.  One could argue that our mpi4py is just a[nother] daos-specific testing tool required, much like al lot else that comes with daos is required.  That we do not publish it is probably the travesty.

> But we don't want to use the system python. AFAIK, the default is python3.6 on all our distros and 3.6 was EOL _**2.5 years ago**_.

But is supported by the distro vendor still, to be clear.

**Comment by brianjmurrell** [comment]
> > Is this test failing in master? It would only be doing that if we'd landed the packaging change already I'd expect.
> 
> Maybe I'm getting it confused with another :)

No, you are not.  This is currently failing in master's daily-testing as a result of https://github.com/daos-stack/daos/commit/dcbaf5505d3d8c6789fb50e3369aad1a2a43316b landing, which had a hack in it that seems to have not worked after the landing.

**Comment by daltonbohning** [comment]
> > This might be okay, but technically it means this test wouldn't work outside our CI environment.
> 
> Yes. This but a raindrop in an ocean of problems with our testing working outside of our CI environment. Although, maybe it's not, entirely. One could argue that our mpi4py is just a[nother] daos-specific testing tool required, much like al lot else that comes with daos is required. That we do not publish it is probably the travesty.

I don't disagree, which is why I'm okay with the hardcoded path if we can't reliably figure it out

> > But we don't want to use the system python. AFAIK, the default is python3.6 on all our distros and 3.6 was EOL _**2.5 years ago**_.
> 
> But is supported by the distro vendor still, to be clear.

But realistically python package maintainers won't always support python 3.6. So we won't get RPMs or even pip packages for python 3.6 package updates. And since there are breaking API changes between 3.6 and the current 3.11/3.12, we can't expect them to.

**Comment by ashleypittman** [comment]
> > > Is this test failing in master? It would only be doing that if we'd landed the packaging change already I'd expect.
> > 
> > 
> > Maybe I'm getting it confused with another :)
> 
> No, you are not. This is currently failing in master's daily-testing as a result of [dcbaf55](https://github.com/daos-stack/daos/commit/dcbaf5505d3d8c6789fb50e3369aad1a2a43316b) landing, which had a hack in it that seems to have not worked after the landing.

Ah, I think maybe it works for el8 which uses python3.6 but fails for leap which uses a different python version?

**Review by ryon-jensen — APPROVED** [review]

**Comment by ryon-jensen** [comment]
Whether workaround or a generic solution, I think we should get this fix in as it does resolve a constant failing issue in daily builds.

File: src/tests/ftest/mpiio/llnl_mpi4py.py
Comment (phender) [inline src/tests/ftest/mpiio/llnl_mpi4py.py:52]: We need a wildcard character added here:
```suggestion
        self.run_test(self.params.get('mpi4py', '/run/test_repo/*'), "mpi4py")
```
So that https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14163/13/artifact/Functional%20on%20EL%208.8/mpiio/llnl_mpi4py.py/job.log will get the value from the test yaml and not use the default:
```
2024-05-06 18:28:48,597 parameters       L0143 DEBUG| PARAMS (key=llnl, path=/run/test_repo/, default=None) => 'testmpio'
```

**Review by phender — CHANGES_REQUESTED** [review]
This PR should also run with:
```
Skip-func-test-el9: false
Skip-func-test-leap15: false
```

**Review by phender — CHANGES_REQUESTED** [review]
-1 so this doesn't land as the change causes the test to fail:
```
2024-05-06 18:29:28,380 process          L0604 INFO | Running '/usr/lib64/mpich/bin/mpirun -genv DAOS_UNS_PREFIX=daos://TestPool_1/TestContainer_1 -hostfile /tmp/hostfile_o6wuc2p3 -np 8 python3.6 /usr/bin/test_io_daos.py'
2024-05-06 18:29:28,780 process          L0416 DEBUG| [stderr] python3.6: can't open file '/usr/bin/test_io_daos.py': [Errno 2] No such file or directory
```
in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14163/13/artifact/Functional%20on%20EL%208.8/mpiio/llnl_mpi4py.py/job.log

We repushed, please also run on EL9 and Leap 15.

**Comment by phender** [comment]
I've landed a fix in https://github.com/daos-stack/daos/commit/e139d7d054f8c6b493a1b3eb736078adb87945bb

**Comment by brianjmurrell** [comment]
> I've landed a fix in [e139d7d](https://github.com/daos-stack/daos/commit/e139d7d054f8c6b493a1b3eb736078adb87945bb)

Is that actually the fix we want vs. this one, which is being driven by [DAOS-15602](https://daosio.atlassian.net/browse/DAOS-15602)?

**Comment by brianjmurrell** [comment]
> [e139d7d](https://github.com/daos-stack/daos/commit/e139d7d054f8c6b493a1b3eb736078adb87945bb) is really just a workaround to get the test passing. We can still move forward with whatever proper solution we need.

Ahh.  OK.  I see.  Perhaps [DAOS-15617](https://daosio.atlassian.net/browse/DAOS-15617) should not be **Resolved** at this point yet then?

In any case, this seems to be passing the `test_mpi4py` test now.  If it gets 2 +1s, I will land daos-stack/mpi4py#12 and then repush this with more appropriate for-landing commit pragmas.

**Review by daltonbohning — APPROVED** [review]

**Comment by phender** [comment]
```
Ahh. OK. I see. Perhaps [DAOS-15617](https://daosio.atlassian.net/browse/DAOS-15617) should not be Resolved at this point yet then?
```
I viewed it as DAOS-15617 was tracking the test failure (which has been worked around) and DAOS-15602 is the ticket to resolve however/wherever we want to install the `test_io_daos.py` python script.

**Comment by phender** [comment]
> Ahh. OK. I see. Perhaps [DAOS-15617](https://daosio.atlassian.net/browse/DAOS-15617) should not be **Resolved** at this point yet then?

I viewed it as DAOS-15617 was tracking the test failure (which has been worked around) and DAOS-15602 is the ticket to resolve however/wherever we want to install the test_io_daos.py python script.

**Review by ryon-jensen — APPROVED** [review]

**Review by phender — COMMENTED** [review]
Code changes look good here, but I have a change request in https://github.com/daos-stack/mpi4py/pull/12

**Comment by brianjmurrell** [comment]
> Code changes look good here, but I have a change request in [daos-stack/mpi4py#12](https://github.com/daos-stack/mpi4py/pull/12)



> Code changes look good here, but I have a change request in [daos-stack/mpi4py#12](https://github.com/daos-stack/mpi4py/pull/12)

@phender Were your concerns on daos-stack/mpi4py#12 resolved?

**Comment by phender** [comment]
We are seeing failures in Leap 15 and EL 9 because https://github.com/daos-stack/mpi4py/pull/12 has already landed w/o this change and they installed (python3-mpi4py-tests-3.1.6-1):
```
avocado.core.exceptions.TestFail: No test repo found in python site-packages
```
- https://build.hpdd.intel.com/job/daos-stack/job/daos/job/daily-testing/179/artifact/Functional%20on%20EL%209/mpiio/llnl_mpi4py.py/job.log
- https://build.hpdd.intel.com/job/daos-stack/job/daos/job/daily-testing/179/artifact/Functional%20on%20Leap%2015.5/mpiio/llnl_mpi4py.py/job.log

(EL8 ran with python3-mpi4py-tests-3.0.3-4)

**Comment by phender** [comment]
All tests passed in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14163/15/testReport/FTEST_mpiio/LlnlMpi4py/
using the python3-mpi4py-tests-3.1.6-1 RPM.

**Pull request merged**
